addons/purchase_stock/models/res_partner.py

```diff
@@ -48,3 +48,35 @@ def _compute_on_time_rate(self):
             on_time, ordered = numbers
             partner.on_time_rate = on_time / ordered * 100 if ordered else -1   # use negative number to indicate no data
         (self - seen_partner).on_time_rate = -1
+
+    @api.model
+    def name_search(self, name='', domain=None, operator='ilike', limit=100):
+        base_res = super().name_search(name, domain, operator, limit)
+        product_id = self.env.context.get('product_id')
+        if not product_id:
+            return base_res
+
+        product = self.env['product.product'].browse(product_id)
+        if not product.seller_ids:
+            return base_res
+
+        unique_sellers = {s.partner_id.id for s in product.seller_ids}
+
+        res_minus_sellers = [partner for partner in base_res if partner[0] not in unique_sellers]
+        product_sellers = [partner for partner in base_res if partner[0] in unique_sellers]  # selection logic is handled in _select_seller
+
+        return product_sellers + res_minus_sellers
+
+    @api.depends('name')
+    def _compute_display_name(self):
+        for rec in self:
+            product_id = rec.env.context.get('product_id')
```

Comment:
if not formatted_display_name in context or not product_id in context or not highlight_supplier in context:
    return super()
... code for formating